### PR TITLE
[Dynamo] Initial deduplication pass impl

### PR DESCRIPTION
**Pull request by mlazos**
This PR implements the deduplication pass (blocked by config currently) for dynamo where identical regions from https://github.com/pytorch/pytorch/pull/141381 are replaced with a common subgraph. 

The two phases of deduplication are explained below.

**Subgraph creation**:
Subgraph creation works by taking one representative region from each region group and creating a subgraph from it, which will then be used to replace all regions in the group. This is implemented by first copying all nodes of the region to the new subgraph and then finding all inputs which are not within the region and creating placeholders for them. For the outputs, all regions in a region group need to be scanned to ensure the largest set of outputs is found, and then an output node is created which returns a tuple of all outputs. 

**Graph replacement**:
To replace each region with the extracted subgraph, the node index in the region and argument index within the node's flattened args and kwargs are recorded once during subgraph creation. This allows us to determine which (external to the region) nodes and in which order these nodes are passed as inputs. For the outputs, getitem nodes are created for each output, and all nodes in the region with external outputs are replaced by the proper getitem node. Finally, all original nodes are erased (there should be no uses of these left in the graph).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141384
* __->__ #141383
* #141382
* #141381

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/graph_deduplication.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+from collections import defaultdict
+from typing import Any, DefaultDict, Dict, Iterable, List, Set, Tuple
+
+import torch.fx
+from torch.utils._pytree import tree_flatten
+
+from .graph_region_tracker import Node, Region
+
+
+def apply_graph_deduplication(output_graph) -> None:  # type: ignore[no-untyped-def]
```
Comment (mlazos) [inline torch/_dynamo/graph_deduplication.py:11]: Add docstring

File: torch/_dynamo/graph_deduplication.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                region,
+                get_subgraph_node,
+                node_ind_arg_inds.keys(),
+                inds_with_external_users,
+                subgraph,
+                subgraph_name,
+            )
+
+
+def _replace_region_with_subgraph(
```
Comment (mlazos) [inline torch/_dynamo/graph_deduplication.py:41]: Don't create if there is mutation n stuff on the input

File: torch/_dynamo/graph_deduplication.py (hunk 1 above)
Comment (zou3519) [reply torch/_dynamo/graph_deduplication.py:11]: still looking for docstring lol

**Review by zou3519 — COMMENTED** [review]
code looks good to me, only comments I had was from our offline discussion:
1) docstrings for the entrypoints
2) we should ban mutations and aliasing as a first pass

**Comment by mlazos** [comment]
> code looks good to me, only comments I had was from our offline discussion:
> 
> 1. docstrings for the entrypoints
> 2. we should ban mutations and aliasing as a first pass

Should be fixed now, added docstring and reused the checking we use for invoke_subgraph.

File: torch/_dynamo/graph_deduplication.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        for arg_ind, in_node in enumerate(flattened_args_kwargs):
+            if in_node not in nodes_unique and isinstance(in_node, Node):
+                external_node_to_indices[in_node].append((node_ind, arg_ind))
+
+    return external_node_to_indices
+
+
+def _get_all_output_indices(regions: List[Region]) -> List[int]:
+    # Scan all regions to get the set of all possible output nodes indices in the region
+    # perhaps it's possible to get this info some other way?
```
[the remark this answers is not in the document]
Comment (mlazos) [reply torch/_dynamo/graph_deduplication.py:130]: I'll update the comment actually, what I was thinking was maybe dynamo or the region tracking could do this more efficiently.

[the remark this answers is not in the document]
Comment (mlazos) [reply torch/_dynamo/graph_deduplication.py]: updated the comment with my thought for future consideration

File: torch/_dynamo/graph_deduplication.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    for node_ind, arg_ind in node_ind_arg_ind:
+        node = region[node_ind]
+        flattened_args_kwargs, _ = tree_flatten((node.args, node.kwargs))
+        sub_args.append(flattened_args_kwargs[arg_ind])
+
+    invoke_args = (get_subgraph_node, subgraph_name, tuple(sub_args))
+    fake_inputs = [node.meta["example_value"] for node in sub_args]
+
+    if has_potential_input_alias_or_mutation(sub_gm, fake_inputs):
+        unimplemented("NYI: replacing subgraph with input alias or mutation")
```
Comment (zou3519) [inline torch/_dynamo/graph_deduplication.py:107]: Why raise unimplemented and then try-catch it? Could we just return here?

File: torch/_dynamo/graph_deduplication.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    sub_args = []
+    for node_ind, arg_ind in node_ind_arg_ind:
+        node = region[node_ind]
+        flattened_args_kwargs, _ = tree_flatten((node.args, node.kwargs))
+        sub_args.append(flattened_args_kwargs[arg_ind])
+
+    invoke_args = (get_subgraph_node, subgraph_name, tuple(sub_args))
+    fake_inputs = [node.meta["example_value"] for node in sub_args]
+
+    if has_potential_input_alias_or_mutation(sub_gm, fake_inputs):
```
Comment (zou3519) [inline torch/_dynamo/graph_deduplication.py:106]: Do you  have a test that a region with input alias / mutation correctly does not get put into an invoke_subgraph?

Comment (mlazos) [reply torch/_dynamo/graph_deduplication.py:106]: No, I'll add

File: test/dynamo/test_graph_deduplication.py [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+            y0 = y + 1
+            x.add_(x0)
+            y.add_(y0)
+            return x.sum() + y.sum()
+
+        def fn(x, y):
+            x0 = torch.sin(x)
+            y0 = torch.cos(y)
+            # o0 = inner_fn(x0, y0)
+            # o1 = inner_fn(x0, o0)
```
Comment (zou3519) [inline test/dynamo/test_graph_deduplication.py:453]: why commented out?

File: test/dynamo/test_graph_deduplication.py [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+
+    def test_input_aliasing(self):
+        def inner_fn(x, y):
+            x0 = x.view(x.size())
+            return x0.view(x.size())
+
+        def inner_fn2(x, y):
+            x = x * 2
+            y = y * 2
+            return x.sum() + y.sum()
```
Comment (zou3519) [inline test/dynamo/test_graph_deduplication.py:527]: Can you do an inner function that is something like:
```
def inner_fn3(x, y):
    return x.sum, y
```
to see if it thinks anything about the "return y" part?

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**